### DISCUSSION TITLE
How implement a specific IActivityExecutionMiddleware  and configure it in Program.cs by Elsa3.0?

### DISCUSSION THREAD

**Discussion by coolshun** [elsa-workflows/elsa-core#5194] [Q&A]
How implement a specific IActivityExecutionMiddleware  and configure it in Program.cs by Elsa3.0?I want to have a valid example code.

**Comment by glime-ai** [comment]
To implement a specific `IActivityExecutionMiddleware` in Elsa 3.0 and configure it in `Program.cs`, you can follow these steps:

1. Create a class that implements the `IActivityExecutionMiddleware` interface.
2. Implement the `InvokeAsync` method to include your custom logic.
3. Register your middleware with the Elsa services in `Program.cs`.

Here's an example of how you might implement a custom middleware:

```csharp
using Elsa.Services;
using Elsa.Services.Models;

public class MyCustomActivityMiddleware : IActivityExecutionMiddleware
{
    public async ValueTask InvokeAsync(ActivityExecutionContext context, Func<ValueTask> next)
    {
        // Your custom pre-activity logic here

        // Continue with the next middleware in the pipeline
        await next();

        // Your custom post-activity logic here
    }
}
```

Then, in your `Program.cs`, you would register your middleware like this:

```csharp
using Microsoft.Extensions.DependencyInjection;
using Elsa.Extensions;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddElsa(elsa =>
{
    // Register other Elsa services
    // ...

    // Register your custom activity execution middleware
    elsa.AddActivityExecutionMiddleware<MyCustomActivityMiddleware>();
});

var app = builder.Build();

// Configure the HTTP request pipeline.
// ...

app.Run();
```

This will add your custom middleware to the Elsa services, and it will be invoked as part of the activity execution pipeline.

Please note that the exact code may vary depending on your specific requirements and the version of Elsa you are using. Make sure to check the official Elsa documentation or the source code for the most up-to-date examples and best practices.

**Comment by coolshun** [comment]
the code above have a error:
Error CS1061: 'IModule' does not contain a definition for 'AddActivityExecutionMiddleware' and cannot find an accessible extension method 'AddActivityExecutionMiddleware' that accepts the first 'IModule' type parameter. (Are you missing a using directive or assembly reference?) @noxe @petedavis @lahma

**Comment by glime-ai** [comment]
To implement a specific `IActivityExecutionMiddleware` and configure it in `Program.cs` for Elsa 3.0, you would typically follow these steps:

1. Implement your custom middleware by creating a class that implements the `IActivityExecutionMiddleware` interface.
2. In your `Program.cs`, you would configure the services to add your custom middleware to the Elsa pipeline.

Here is an example of how you might implement and configure a custom `IActivityExecutionMiddleware`:

```csharp
using Elsa.Workflows.Contracts;
using Elsa.Workflows.Pipelines.ActivityExecution;

namespace YourNamespace
{
    public class CustomActivityExecutionMiddleware : IActivityExecutionMiddleware
    {
        public async ValueTask InvokeAsync(ActivityExecutionContext context, Func<ValueTask> next)
        {
            // Your custom pre-execution logic here

            // Continue with the pipeline
            await next();

            // Your custom post-execution logic here
        }
    }
}

// In Program.cs
using YourNamespace;

var builder = WebApplication.CreateBuilder(args);

// Add Elsa services and your custom middleware
builder.Services.AddElsa(elsa =>
{
    // ... other Elsa configurations

    // Add your custom middleware to the pipeline
    elsa.UseActivityExecutionMiddleware<CustomActivityExecutionMiddleware>();
});

// ... rest of the Program.cs code
```

Please note that the exact method to add the middleware to the pipeline (`UseActivityExecutionMiddleware`) may vary based on the version of Elsa you are using and the available extension methods. If the method `UseActivityExecutionMiddleware` does not exist, you may need to look for an alternative way to add your middleware to the pipeline, such as using `AddActivityExecutionMiddleware` if it's available.

If you encounter any issues or the method names have changed in the version you are using, you should refer to the official Elsa documentation or seek help from the Elsa community for the most up-to-date information.
